### PR TITLE
feat(list): Specify the element types in the `calciteListOrderChange` event detail.

### PR DESCRIPTION
**Pull request by driskull**
**Related Issue:** #8111

## Summary

- More specifically types the elements within the drag event

### CODE REVIEW

File: packages/calcite-components/src/components/list/list.tsx [hunk 1]
```diff
@@ -800,9 +806,9 @@ export class List implements InteractiveComponent, LoadableComponent, SortableCo
     this.connectObserver();
 
     this.calciteListOrderChange.emit({
-      dragEl: sortItem,
-      fromEl: parentEl,
-      toEl: parentEl,
+      dragEl: sortItem as HTMLCalciteListItemElement,
```
Comment (jcfranco) [inline packages/calcite-components/src/components/list/list.tsx:809]: I think you can avoid type casting here if you type `parentEl` when it's assigned.

File: packages/calcite-components/src/components/list/list.tsx [hunk 2]
```diff
@@ -35,6 +35,12 @@ import {
 } from "../../utils/sortableComponent";
 import { SLOTS as STACK_SLOTS } from "../stack/resources";
 
+type ListDragDetail = DragDetail & {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/list/list.tsx:38]: WDYT about enhancing `DragDetail` to use generics to define the to, from, drag element types? Something like:

```ts
export interface DragDetail<
  To extends HTMLElement = HTMLElement, 
  From extends HTMLElement = HTMLElement, 
  Drag extends HTMLElement = HTMLElement
> {
  toEl: To;
  fromEl: From;
  dragEl: Drag;
  newIndex: number;
  oldIndex: number;
}
```

**Comment by jcfranco** [comment]
Should this PR be using `feat` instead of `fix`? The associated issue and the changes in this PR seem more related to a feature.

File: packages/calcite-components/src/utils/sortableComponent.ts [hunk 3]
```diff
@@ -1,10 +1,14 @@
 import Sortable from "sortablejs";
 const sortableComponentSet = new Set<SortableComponent>();
 
-export interface DragDetail {
-  toEl: HTMLElement;
-  fromEl: HTMLElement;
-  dragEl: HTMLElement;
+export interface DragDetail<
```
Comment (driskull) [inline packages/calcite-components/src/utils/sortableComponent.ts:4]: @jcfranco is this what you were thinking? Will this type the elements correctly?

Comment (jcfranco) [reply packages/calcite-components/src/utils/sortableComponent.ts:4]: It should if you provide the types in the emitter type.

```ts
@Event({ cancelable: false }) calciteListOrderChange: EventEmitter<
  DragDetail<
    HTMLCalciteListElement, 
    HTMLCalciteListElement, 
    HTMLCalciteListItemElement
  >
>;
```

File: packages/calcite-components/src/components/list/list.tsx (hunk 2 above)
Comment (driskull) [reply packages/calcite-components/src/components/list/list.tsx]: @jcfranco I'm seeing this error in the build:

```console
Error: angular-workspace:build: projects/component-library/src/lib/stencil-generated/components.ts:1410:52 - error TS2315: Type 'DragDetail' is not generic.
```

Comment (jcfranco) [reply packages/calcite-components/src/components/list/list.tsx]: It looks like a Stencil 2.x issue (see https://github.com/ionic-team/stencil-ds-output-targets/issues/314). We could test w/ v4 to see if the issue is fixed. If v4 doesn't look any different, we could proceed with your previous approach and revisit the interface once generics are supported.

Comment (driskull) [reply packages/calcite-components/src/components/list/list.tsx:38]: > proceed with your previous approach and revisit the interface once generics are supported.

This sounds like a good approach. We can try updating it when 4x is installed.

File: packages/calcite-components/src/components/list/list.tsx [hunk 4]
```diff
@@ -800,9 +800,9 @@ export class List implements InteractiveComponent, LoadableComponent, SortableCo
     this.connectObserver();
 
     this.calciteListOrderChange.emit({
-      dragEl: sortItem,
-      fromEl: parentEl,
-      toEl: parentEl,
+      dragEl: sortItem as HTMLCalciteListItemElement,
```
Comment (jcfranco) [inline packages/calcite-components/src/components/list/list.tsx:803]: Do you need to type cast here? I think all of the elements are properly typed to what the interface expects.

**Review by jcfranco — APPROVED** [review]
✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨
✨📝✨✨✨📝✨📝📝📝✨✨📝📝📝✨📝📝📝📝✨📝✨
✨📝📝✨✨📝✨✨📝✨✨📝✨✨✨✨📝✨✨✨✨📝✨
✨📝✨📝✨📝✨✨📝✨✨📝✨✨✨✨📝📝📝✨✨📝✨
✨📝✨✨📝📝✨✨📝✨✨📝✨✨✨✨📝✨✨✨✨✨✨
✨📝✨✨✨📝✨📝📝📝✨✨📝📝📝✨📝📝📝📝✨📝✨
✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨✨

**Comment by driskull** [comment]
@jcfranco still getting an angular error:

```
Error: angular-workspace:build: projects/component-library/src/lib/stencil-generated/components.ts:1396:15 - error TS2305: Module '"@esri/calcite-components"' has no exported member 'ListDragDetail'.
angular-workspace:build: 
angular-workspace:build: 1396 import type { ListDragDetail as ICalciteListListDragDetail } from '@esri/calcite-components';
`

**Comment by jcfranco** [comment]
@benelan Have you seen the above error before? ☝️

**Pull request merged**